### PR TITLE
DataFusion + Conbench Integration

### PR DESCRIPTION
**Pull request by dianaclarke**
Here's a minimal DataFusion + Conbench[1] proof of concept.

[1] https://github.com/conbench/conbench

A few notes (areas for improvement, caveats, etc):

 - Criterion results are in nanoseconds, but the smallest unit
Conbench currently speaks is seconds (because Conbench was initially
for macro not micro benchmarking). I suspect most places in Conbench
would work just fine if nanoseconds were passed in, but I need to
audit the code for any places that assume seconds if it isn't a
throughput benchmark.

- If the Criterion benchmarks were named better, I could tag them
better in Conbench. For example, I suspect sqrt_20_12, sqrt_20_9,
sqrt_22_12, and sqrt_22_14 are parameterized variations of the same
benchmark, and if they were named something like "sqrt, foo=20,
bar=12", I could batch them together & tag their parameters so that
Conbench would automatically graph them in relation to each other. I
was sort of able to do this with the following benchmarks (because
there was a machine readable pattern). Anyhoo, that's easy enough to
do down the road as a last integration step, and it does appear from
the Criterion docs that they have their own recommendations for how to
do this.

    - window partition by, u64_narrow, aggregate functions
    - window partition by, u64_narrow, built-in functions
    - window partition by, u64_wide, aggregate functions
    - window partition by, u64_wide, built-in functions

- While Criterion benchmarks can also measure throughput in some
cases, all the arrow-datafusion benchmarks were in elapsed time (not
sure about the arrow-rs benchmarks), so I didn't bother writing code
to support potential throughput results from
arrow-datafusion/arrow-rs, but we may need to revisit that.

- We probably want to add some additional context, like the
arrow-rs/arrow-datafusion version, rust version, any compiler flags,
etc.

### CODE REVIEW

File: conbench/.flake8 [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (dianaclarke) [inline conbench/.flake8:1]: From a Python perspective, it's probably a mistake to name this directory `conbench`. Thoughts on what you would like `arrow-datafusion/conbench/` to be named?

- `arrow-datafusion/_conbench/`?
- `arrow-datafusion/conbench-benchmarks/`?

File: conbench/_criterion.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    # time-per-iteration To calculate the time-per-iteration, use
+    # sample_measured_value/iteration_count
+    # -- https://bheisler.github.io/criterion.rs/book/user_guide/csv_output.html
+    count = int(row["iteration_count"])
+    sample = float(row["sample_measured_value"])
+    return sample / count / 10**9
+
+
+def _parse_benchmark_group(row):
+    parts = row["group"].split(",")
```
Comment (dianaclarke) [inline conbench/_criterion.py:40]: TODO: strip extra spaces, or abort trying to bucket and just go with the names as-is?

```
    "tags": {
        "name": " u64_wide,  built-in functions",
        "suite": "window partition and order by"
    },
```

**Comment by alamb** [comment]
Thank you @dianaclarke  -- I will try and review this carefully over the next day or two. CC @andygrove  and @Dandandan

**Comment by dianaclarke** [comment]
> Thank you @dianaclarke -- I will try and review this carefully over the next day or two. CC @andygrove and @Dandandan

No rush & no pressure to merge this (or the sister `arrow-rs` pull request). I don't actually work on anything Arrow related anymore – just didn't want to leave you hanging. Cheers

**Review by andygrove — APPROVED** [review]
LGTM. I followed the instructions and conbench produced reports showing the benchmark results.

**Comment by andygrove** [comment]
If there are no objections, I plan to merge this PR next week so that I can proceed to the next step of integrating this with CI. The changes in this PR are self-contained so I think it is low risk to merge.

**Review by Dandandan — APPROVED** [review]
LGTM

**Comment by houqp** [comment]
Thank you @dianaclarke , this is going to be a big productivity boost for us on all performance related PRs :)

**Pull request merged**